Feature/add files to gist

g14a opened the pull request:
Adds functionality to add files to an existing gist. Add files with absolute path, relative path, files present in the current working directory and new files from stdin.

Closes #2903

g14a committed:
Feature of adding new files to an existing Github gist

g14a committed:
Add existing files in the current wd to gist

g14a committed:
Add files by absolute path to gist

g14a committed:
Add test cases and improve errors with color schemes

g14a committed:
Merge branch 'trunk' of https://github.com/cli/cli into feature/add-files-to-gist

g14a committed:
Remove unwanted prompt for user. Unwanted test as well

g14a committed:
gofmt

g14a committed:
add go lint to pass checks

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -44,11 +48,18 @@ func NewCmdEdit(f *cmdutil.Factory, runF func(*EditOptions) error) *cobra.Comman
 	cmd := &cobra.Command{
 		Use:   "edit {<id> | <url>}",
-		Short: "Edit one of your gists",
+		Short: "Edit one of your gists or add new ones to it",
```
```suggestion
		Short: "Edit or add files in a gist",
```

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -222,3 +297,36 @@ func updateGist(apiClient *api.Client, hostname string, gist *shared.Gist) error
 	return nil
 }
+
+func processFiles(filename string) (string, bool, error) {
```
It seems the purpose of this function is to determine whether or not a file exists, so it should be called `fileExists`

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -222,3 +297,36 @@ func updateGist(apiClient *api.Client, hostname string, gist *shared.Gist) error
+		}
+	}
+
+	return filename, fileExists, nil
```
`filename` is not modified and does not need to be returned here

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -222,3 +297,36 @@ func updateGist(apiClient *api.Client, hostname string, gist *shared.Gist) error
+			fileExists = true
+		}
+	}
+
```
why continue execution, here? In my testing `os.Stat` was sufficient for finding the various file paths i threw at this.

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -222,3 +297,36 @@ func updateGist(apiClient *api.Client, hostname string, gist *shared.Gist) error
+func processFiles(filename string) (string, bool, error) {
+	fileExists := false
+
+	if fi, err := os.Stat(filename); err != nil {
+	} else {
```
this conditional is a no-op. should this be an early return? I'm not following why, if this failed, the later glob would help.

g14a replied:
Yes you're right. I guess I over thought this. The above one too.

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -99,89 +111,152 @@ func editRun(opts *EditOptions) error {
-		if len(candidates) == 1 {
-			break
-		}
+	} else {
```
this conditional is getting out of hand. I'd like to see helper functions and early exits for adding and editing files.

g14a committed:
Resolved PR review comments and test cases

g14a committed:
Merge branch 'trunk' of https://github.com/cli/cli into feature/add-files-to-gist

g14a committed:
Add go fmt to pass ci/cd

rawlingsluke19 commented:
On Thu, Feb 18, 2564 BE at 10:33 PM Gowtham Munukutla <
notifications@github.com> wrote:

> *@g14a* commented on this pull request.
> ------------------------------
>
> In pkg/cmd/gist/edit/edit.go
> <https://github.com/cli/cli/pull/2997#discussion_r578919330>:
>
> > +	if fi, err := os.Stat(filename); err != nil {
> +	} else {
>
> Yes you're right. I guess I over thought this. The above one too.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cli/cli/pull/2997#discussion_r578919330>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASBANNGQ3WKFWQEHGN3E4YTS7XSZDANCNFSM4X2VDNNA>
> .
>
-- 
Luke Rawlings

g14a commented:
@vilmibm I've resolved your comments and updated the PR.

g14a commented:
Anything else to be done? @vilmibm

g14a committed:
Merge branch 'trunk' of https://github.com/cli/cli into feature/add-files-to-gist

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -222,3 +259,74 @@ func updateGist(apiClient *api.Client, hostname string, gist *shared.Gist) error
+}
+
+func fileExists(filename string) (bool, error) {
+
```
```suggestion
```

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -222,3 +259,74 @@ func updateGist(apiClient *api.Client, hostname string, gist *shared.Gist) error
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
```
```suggestion
		}
		return false, err
```

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -222,3 +259,74 @@ func updateGist(apiClient *api.Client, hostname string, gist *shared.Gist) error
+			Content:  string(content),
+		}
+		return filesToAdd, nil
+	} else {
```
I've been playing around with this in my terminal and let's please go with my original request of not opening the editor to add a new file and instead just printing an error.

If a user mistypes a filename they are suddenly dropped into an editor which is very confusing and I don't think it's too much to ask that a user have a file already created that they want to add.

g14a replied:
oh you're right. I haven't tested that case. Makes sense. 👍🏼

vilmibm commented on pkg/cmd/gist/edit/edit.go:
```diff
@@ -99,8 +111,33 @@ func editRun(opts *EditOptions) error {
+		if err != nil {
+			return err
+		}
+
+		fmt.Printf("%v", files)
```
```suggestion
```

vilmibm reviewed (dismissed):
Looking much better and almost there!

I left some suggestions and my final request is to remove the ability to add new files when a specified file doesn't exist.

g14a committed:
updating tests WIP

g14a committed:
Merge branch 'trunk' of https://github.com/cli/cli into feature/add-files-to-gist

g14a committed:
remove unwanted tests and unwanted functionality

g14a committed:
add fixturefile const in tests

g14a commented:
Followed your suggestions @vilmibm and updated the PR.

g14a committed:
rebase with trunk

g14a commented:
Anything else to be done here? @vilmibm

mislav committed:
:nail_care: cleanup `gist edit -a` feature

mislav committed:
Improve error handling and avoid writing confirmation to stdout

Right now the `gist edit` command doesn't write anything to stdout, so
let's keep it that way until we want to intentionally provide some
feedback in the terminal.

mislav reviewed (approved):
Thank you for all your work!

This looks good to me. I've pushed some polish changes 👍
